<?xml version="1.0" encoding="UTF-8"?>

<launch>
  <!-- ========== Arguments ========== -->
  <!-- Dataset -->
  <arg name="use_doals" default="false" />  <!-- Which dataset to play -->
  <arg name="bag_file" default="/media/dylan/Dylan_1TB/CMU_Research/Data/preprocess_humans.bag" />  <!-- Full path to the bag file to play -->
  <arg name="player_rate" default="1" />  <!-- Real time rate of bag being played -->
  
  <!-- Drift Simulation -->
  <arg name="drift_simulation_rollout" default="" />  <!-- Leave empty to ignore drift or specify a rollout that matches the dataset bein played --> 
  
  <!-- Evaluation -->
  <arg name="evaluate" default="false" />  <!-- Whether to save evaluation data -->  
  <arg name="eval_output_path" default="/home/$(env USER)/dynablox_output" />  <!-- Where to save evaluation data  -->
  <arg name="ground_truth_file" default="/home/$(env USER)/data/DOALS/hauptgebaeude/sequence_1/indices.csv" />  <!-- GT data file. Currently supports DOALS -->
  
  <!-- Motion Detector -->
  <arg name="config_file" default="motion_detector/default.yaml" />  <!-- Configuration of Dynablox -->
  <arg name="visualize" default="true" />  <!-- Whether to display RVIZ visualizations -->
  
  <!-- Cluster Extractor -->
  <arg name="max_cluster_topics" default="20"/>
  <arg name="run_cluster_extractor" default="true"/>
  <arg name="cluster_output_dir" default="/home/dylan/dynablox/cluster_data"/>
  <arg name="cluster_file_prefix" default="cluster"/>
  <arg name="save_clusters_as_bin" default="true"/>
  
  <!-- ========== Run Nodes ========== -->
  <!-- Auto-generated arguments -->
  <arg name="use_drift" value="$(eval arg('drift_simulation_rollout')!='')" />

  <!-- Play the data -->
  <include file="$(find dynablox_ros)/launch/play_doals_data.launch" pass_all_args="true" if="$(arg use_doals)"/>   
  <include file="$(find dynablox_ros)/launch/play_dynablox_data.launch" pass_all_args="true" unless="$(arg use_doals)"/> 
   
  <!-- Drift Simulation -->
  <node name="drift_reader" pkg="drift_simulation" type="drift_reader" output="screen" args="--alsologtostderr" if="$(arg use_drift)">
    <param name="drift_data_file_name" value="$(find drift_simulation)/config/rollouts/$(arg drift_simulation_rollout)" />
  </node>

  <!-- Motion Detection -->
  <node name="motion_detector" pkg="dynablox_ros" type="motion_detector" output="screen" args="--alsologtostderr" required="true">
    <remap from="pointcloud" to="/ouster/points" />
    <remap from="pointcloud" to="/pointcloud_drifted" if="$(arg use_drift)" />

    <!-- config -->
    <rosparam command="load" file="$(find dynablox_ros)/config/$(arg config_file)" />

    <!-- evaluation -->
    <param name="evaluation/ground_truth/file_path" value="$(arg ground_truth_file)" />
    <param name="evaluation/output_directory" value="$(arg eval_output_path)" />
    <param name="evaluate" value="$(arg evaluate)" />    

    <!-- max_cluster_topics parameter -->
    <param name="max_cluster_topics" value="$(arg max_cluster_topics)"/>
  </node>
  
  <!-- Use Cluster Extractor if set to True -->
  <node pkg="dynablox_ros" type="cluster_extractor" name="cluster_extractor" output="screen" if="$(arg run_cluster_extractor)">
    <param name="output_dir" value="$(arg cluster_output_dir)"/>
    <param name="file_prefix" value="$(arg cluster_file_prefix)"/>
    <param name="save_as_bin" value="$(arg save_clusters_as_bin)"/>
    <param name="num_clusters" value="$(arg max_cluster_topics)"/>
  </node>
  
  <!-- Visualize -->
  <include file="$(find dynablox_ros)/launch/visualizer.launch" if="$(arg visualize)"/> 

</launch>
